Add world with moving platform

mbjd opened the pull request:
Same as default world, but with a flat platform to simulate shipboard takeoff & landing. 

It will not move on its own. Velocity can be set by publishing a `gz::msgs::Twist` message to `/model/flat_platform/link/platform_link/cmd_vel`, containing velocity and angular velocity. The platform uses the [`gz::sim::systems::VelocityControl`](https://github.com/gazebosim/gz-sim/blob/gz-sim7/src/systems/velocity_control/VelocityControl.cc) plugin to move when receiving these messages. Working example [here](https://github.com/PX4/PX4-Autopilot/compare/pr-fw_ctrl_api...pr-fw_ctrl_api-gazebo_platform).

(the CI check complains that it is not a model SDF file. It's a world SDF file and as such does not need a `model` top level element as checked in [check_sdf.cc](https://github.com/PX4/PX4-gazebo-models/blob/main/sdf_parsing/check_sdf.cc). It is also a valid SDF file in the first place, otherwise the check would exit even earlier and output `worlds/moving_platform.sdf is not a valid SDF file!`. Should the check be disabled for worlds, or extended to handle worlds correctly?)

perrrre reviewed (commented):
If possible I would import the default world instead of copying, like we did with [x500 gimbal](https://github.com/PX4/PX4-gazebo-models/blob/main/models/x500_gimbal/model.sdf#L5)

mbjd commented:
@perrrre I thought about that, would be very neat, but the [SDF format spec](http://sdformat.org/spec?ver=1.9&elem=world) says: 

    Included resources can only contain one 'model', 'light' or 'actor' element.

What do you think about the following alternative? We load the default world, and have a separate platform.sdf containing just the model. Then we dynamically add it from GZBridge, switched on or off with an extra environment variable.

This is essentially what I had implemented previously https://github.com/PX4/PX4-Autopilot/commit/adabdf42d05044651badab00d2b34ce1662d776a -- I considered it not so clean because we add an extra environment variable and more logic in GZBridge, whereas the current version only relies on specifying the right world. But I see that repeating the default world is suboptimal too. If you think that weighs more, I'm happy to revert the related changes.

dakejahl commented:
This should probably be a custom plugin. The GZBridge is meant to be a simple uORB <--> gz::transport bridge. 

I've been working on implementing custom plugins architecture, I'll get it merged today.
https://github.com/PX4/PX4-Autopilot/pull/24153

I've also got a PR which refactors the GZBridge quite significantly. I will also merge it today.
https://github.com/PX4/PX4-Autopilot/pull/24421

So my thinking is that this should be a new world that includes a "moving platform" plugin and your platform model.sdf. The plugin would be responsible for driving around your model.

dakejahl commented:
Can you take a look at this discussion from Ardupilot? Adding an actual ship/ocean would be the optimal way to do this
https://discuss.ardupilot.org/t/ship-landing-ignition-simulation/80663/35?page=2

plugin repo
https://github.com/srmainwaring/asv_wave_sim/tree/prs/demo/pr-ship-landing-v2-rebase

mbjd commented:
@dakejahl thanks for the input, and those perfectly timed merges! I agree a plugin is probably the way to go, will implement it that way.

dakejahl commented:
See here how to add plugins (no longer in world files)
https://github.com/PX4/PX4-Autopilot/pull/24441/files#diff-6e22639fa95bcd309a2b1a8754533ebb752b0fa06d08480ed6d516a5fbb97b9eR11

mbjd commented:
Marking this as draft while I get the plugin implementation sorted and wait for https://github.com/PX4/PX4-Autopilot/pull/24441 to merge.

dakejahl commented:
I updated the PR with a plugin template, if you do go down the custom plugin route.

@mbjd did you look at https://github.com/srmainwaring/asv_wave_sim? I think this would work perfectly. It will also help out others who are working on boats. 

@patrickelectric any interest in seeing a gz ocean world?

mbjd commented:
@dakejahl very cool, thanks for the plugin template, will start once merged. 

About the wave sim -- we are actually targeting generic moving platforms, not just boats specifically. As such we don't need/want very realistic boat movement and are fine with a simplistic flat plate, driven by low-pass filtered white noise. Thanks though for the pointer to the ardupilot discussion, that definitely has some helpful bits of info!

patrickelectric commented:
> I updated the PR with a plugin template, if you do go down the custom plugin route.
> 
> @mbjd did you look at [srmainwaring/asv_wave_sim](https://github.com/srmainwaring/asv_wave_sim?rgh-link-date=2025-03-05T00%3A28%3A31Z)? I think this would work perfectly. It will also help out others who are working on boats.
> 
> @patrickelectric any interest in seeing a gz ocean world?

That would be great!

mbjd commented:
Un-drafting this again. Changes:
 - separated model into its own SDF file
 - the velocity controller is now a gazebo plugin rather than ad-hoc in PX4. Correspondingly we include that plugin and the velocity control plugin in [models/moving_platform/model.sdf](https://github.com/PX4/PX4-gazebo-models/pull/83/files#diff-29a701700bff98056fe3f7606edd4a0d50bd20849db61534a0e0e209d22d2212)

Still the default world is essentially repeated, the only difference being `<include> <uri>model://moving_platform</uri> </include>`. If someone knows a way to ditch this repetition that would be cool but the other worlds seem to do it this way too. 

The plugin implementation on the PX4 side is working now, based on the brand new plugin infrastructure: https://github.com/PX4/PX4-Autopilot/tree/gz_moving_platform

perrrre reviewed (approved):
LGTM

Jaeyoung-Lim commented:
@mbjd I don't think this should have been merged. Please address the changes suggested by @dakejahl

mbjd commented on models/moving_platform/model.sdf:
```diff
@@ -0,0 +1,48 @@
+      </sensor>
+    </link>
+    <!-- plugin for sending direct velocity commands -->
+    <plugin
+      filename="gz-sim-velocity-control-system"
+      name="gz::sim::systems::VelocityControl">
+    </plugin>
+    <plugin
+      filename="libMovingPlatformController.so"
+      name="custom::MovingPlatformController">
+    </plugin>
```
Plugins *in general* or world plugins? ATM this is a model plugin, binding to the moving platform. Happy to change it though and add it via server.config if there is a concrete benefit. 

In my tests it worked, why shouldn't it? PX4 branch https://github.com/PX4/PX4-Autopilot/tree/gz_moving_platform, start with: ` PX4_GZ_MODEL_POSE=0,0,2.2,0,0,0 PX4_GZ_WORLD=moving_platform make px4_sitl gz_standard_vtol`

mbjd commented:
@Jaeyoung-Lim yes, my apologies, should have waited for all reviews. Will make followup PR if this changes.
